calico/openstack/mech_calico.py

```diff
@@ -196,14 +206,148 @@ def _init_state(self):
+                    break
+
+                self._handle_status_update(response)
+                # Since we're polling on a subtree, we can't just
+                # increment the index, we have to look at the
+                # modifiedIndex.
+                self._next_etcd_index = max(self._next_etcd_index,
+                                           int(response.modifiedIndex)) + 1
+        except (ReadTimeoutError, SocketTimeout) as e:
+            # This is expected when we're doing a poll and nothing
+            # happened. socket timeout doesn't seem to be caught by
+            # urllib3 1.7.1. Simply reconnect.
+            LOG.debug("Read from etcd timed out (%r), retrying.", e)
+        except (etcd.EtcdClusterIdChanged, etcd.EtcdEventIndexCleared) as e:
+            # Complete resync is needed.
+            LOG.warning("Out of sync with etcd (%r).  Reconnecting "
+                     "for full sync.", e)
+            try:
+                context = ctx.get_admin_context()
+                self.complete_resync(context)
+            except Exception:
+                LOG.exception("Error in full resync.")
+        except Exception as e:
+            LOG.info("Handling status update failed. ({})".format(e))
+
+    def _handle_status_update(self, response):
+        """
+        Handle updates of felix instances written to etcd. Pass status
+        updates to neutron database.
+
+        :param response: update written to etcd by one of felix instances
+        """
+        # Only get info from uptime keys - since those have ttl, we know
+        # if key was updated or created.
+        key = response.key
+        if not is_uptime_key(key):
+            return
+        # Ignore deleted and expired keys - their deletion is noticed by
+        # other timeout algorithm before reaching the horizon UI.
+        if response.action == "delete" or response.action == "expire":
+            return
+
+
+        hostname = _hostname_from_status_key(key)
+        agent_state = {'agent_type': AGENT_TYPE_FELIX,
+                       'binary': '',
+                       'host': hostname,
+                       'topic': constants.L2_AGENT_TOPIC}
+        if response.newKey:
+            agent_state['start_flag'] = True
+
+        self.db.create_or_update_agent(self._db_context, agent_state)
+
+
+    def _register_initial_felixes(self):
+        '''
+        Read through etcd status subtree and pass agent update of hosts which
+        have an uptime key (i.e. felix is heartbeating).
+        '''
+        response = self._client.read(STATUS_DIR, recursive=True)
+
+        # Read through hosts
+        for host in response._children:
+            # Read through status updates of the node.
+            for status in host['nodes']:
+                key = status['key']
+                # If host has uptime key, we pass the host to the neutron
+                # database with start_flag set True.
+                if key[-7:] == '/uptime':
```

Comment:
Not keen on the magic number -7 here and this will break if we add any other keys that happen to end in uptime.  Better to parse this properly (probably with a regex in the datamodel module).